### PR TITLE
Mobile 4047

### PR DESCRIPTION
**Pull request by crazyserver**

### CODE REVIEW

File: src/testing/services/behat-runtime.ts [hunk 1]
```diff
@@ -348,23 +352,18 @@ export class TestingBehatRuntimeService {
...
 
         try {
-            // TODO We should generalize this to work with other pages. It's not possible to use
-            // an IonRefresher instance because it doesn't expose any methods to trigger refresh,
-            // so we'll have to find another way.
-
-            const dashboard = this.getAngularInstance<CoreCoursesDashboardPage>(
-                'page-core-courses-dashboard',
-                'CoreCoursesDashboardPage',
+            const ionRefresher = this.getAngularInstance<IonRefresher>(
```
Comment (NoelDeMartin) [inline src/testing/services/behat-runtime.ts:355]:
```suggestion
            // 'el' is protected, but there's no other way to trigger refresh programatically.
            const ionRefresher = this.getAngularInstance<{ el: HTMLIonRefresherElement }>(
```

File: src/testing/services/behat-runtime.ts [hunk 2]
```diff
@@ -348,23 +352,18 @@ export class TestingBehatRuntimeService {
...
+            if (!ionRefresher) {
+                return 'ERROR: It\'s not possible to pull to refresh the current page.';
             }
 
-            await new Promise(resolve => {
-                dashboard.refreshDashboard({ complete: resolve } as IonRefresher);
-            });
+            // el is protected but there's no other way.
+            const refresher: HTMLIonRefresherElement = (ionRefresher as any).el;
+            refresher.dispatchEvent(new CustomEvent('ionRefresh'));
```
Comment (NoelDeMartin) [inline src/testing/services/behat-runtime.ts:366]:
```suggestion
            ionRefresher.el.dispatchEvent(new CustomEvent('ionRefresh'));
```

**Review by NoelDeMartin — COMMENTED** [review]
I didn't review the entire PR (I think @dpalou is already doing that, looking at the status of the [jira issue](https://tracker.moodle.org/browse/MOBILE-4047)), but I spotted some improvements.

File: src/testing/services/behat-runtime.ts (hunk 1 above)
Comment (crazyserver) [reply src/testing/services/behat-runtime.ts:355]: Thanks changed

File: src/testing/services/behat-runtime.ts (hunk 2 above)
Comment (crazyserver) [reply src/testing/services/behat-runtime.ts:366]: Thanks for your suggestions! Changed!

**Review by dpalou — CHANGES_REQUESTED** [review]

File: local_moodleappbehat/tests/behat/behat_app.php [hunk 3]
```diff
@@ -841,9 +841,33 @@ public function i_close_the_browser_tab_opened_by_the_app() {
...
      */
     public function i_switch_offline_mode(string $offline) {
-        $this->runtime_js("network.setForceOffline($offline)");
+        $this->i_switch_network_connection($offline == 'true' ? 'offline' : 'online');
+    }
+
+    /**
+     * Switch network connection.
+     *
+     * @When /^I switch network connection to (online|wifi|cellular|offline)$/
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply local_moodleappbehat/tests/behat/behat_app.php:860]: Ok! Done!

File: src/core/features/comments/tests/behat/basic_usage.feature [hunk 4]
```diff
@@ -36,8 +36,8 @@ Feature: Test basic usage of comments in app
         | Field description | Test field description |
     And I press "Save"
     And I close the browser tab opened by the app
-    When I entered the course "Course 1" as "teacher1" in the app
-    And I press "Data" in the app
+
+    When I pull to refresh in the app
```
[the remark this answers is not in the document]
Comment (crazyserver) [reply src/core/features/comments/tests/behat/basic_usage.feature:40]: I changed courses and choice ones ;-)

**Review by dpalou — APPROVED** [review]

**Pull request merged**